src/main/java/co/cask/hydrator/plugin/source/batch/SalesforceRecordReader.java

```diff
@@ -0,0 +1,134 @@
+import com.sforce.async.BulkConnection;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.StringReader;
+
+/**
+ * RecordReader implementation, which reads a single salesforce batch from bulk job
+ * provided in InputSplit
+ */
+public class SalesforceRecordReader extends RecordReader<String, String> {
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceRecordReader.class);
+
+  private BulkConnection bulkConnection;
+  private String jobId;
+  private String batchId;
+  private BufferedReader queryReader = null;
+
+  private String key;
+  private String value;
+
+  /**
+   * Get csv from a single salesforce batch
+   *
+   * @param inputSplit specifies batch details
+   * @param taskAttemptContext task context
+   * @throws IOException can be due error during reading query
+   * @throws InterruptedException interrupted sleep while waiting for batch results
+   */
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
+    throws IOException, InterruptedException {
+
+    SalesforceSplit salesforceSplit = (SalesforceSplit) inputSplit;
+    jobId = salesforceSplit.getJobId();
+    batchId = salesforceSplit.getBatchId();
+
+    Configuration conf = taskAttemptContext.getConfiguration();
+    SalesforceBatchSource.Config pluginConfig = new SalesforceBatchSource.Config(conf);
+
+    try {
+      AuthenticatorCredentials credentials = pluginConfig.getAuthenticatorCredentials();
+      bulkConnection = new BulkConnection(Authenticator.createConnectorConfig(credentials));
+      String queryResponse = SalesforceBulkUtil.waitForBatchResults(bulkConnection, jobId, batchId);
+
+      queryReader = new BufferedReader(new StringReader(queryResponse));
+      key = queryReader.readLine(); // first line of csv contains names of columns
+    } catch (AsyncApiException e) {
+      throw new RuntimeException("Exception while communicating with Salesforce API", e);
+    }
+  }
+
+  /**
+   * Reads one row from csv. Sometimes the entries in csv can be multiline.
+   * That's why one row does not equal one text line.
+   *
+   * @return returns false if no more data to read
+   * @throws IOException exception from readLine from query csv
+   */
+  @Override
+  public boolean nextKeyValue() throws IOException {
+    StringBuilder result = new StringBuilder();
+
+    for (;;) {
+      String line = queryReader.readLine();
+
+      // end of response for this record reader
+      if (line == null) {
+        return false;
+      }
+
+      result.append(line);
+
+      // All values in salesforce csv always start are enquoted with ". That's how we know that the line break is end
```

Comment:
```suggestion
      // All values in Salesforce csv always start with double-quote. That's how we know that the line break is end
```

Reply:
not doulbe-quote. But a single.

Reply:
Could you please check again, generally the reference is the following:
```
SINGLE_QUOTE: '
DOUBLE_QUOTE: "
```

Reply:
oh. That's what you mean. Got it.